// @backstage/plugin-search/dist/esm/index-0a13cd80.esm.js
import React from 'react';
import SearchIcon from '@material-ui/icons/Search';
import { SidebarItem } from '@backstage/core-components';
import { a as SearchModalProvider, u as useSearchModal, b as SearchModal } from './index-0387f230.esm.js';
import '@material-ui/core';
import '@material-ui/icons/Launch';
import '@material-ui/core/styles';
import '@backstage/plugin-search-react';
import '@backstage/core-plugin-api';
import '@backstage/errors';
import 'qs';
import '@backstage/version-bridge';
import 'react-use/lib/usePrevious';
import 'react-router';
import 'react-use/lib/useEffectOnce';
import '@material-ui/icons/ExpandMore';
import '@material-ui/icons/FontDownload';
import 'react-router-dom';

const SidebarSearchModalContent = (props) => {
  const { state, toggleModal } = useSearchModal();
  const Icon = props.icon ? props.icon : SearchIcon;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(SidebarItem, {
    className: "search-icon",
    icon: Icon,
    text: "Search",
    onClick: toggleModal
  }), /* @__PURE__ */ React.createElement(SearchModal, {
    ...state,
    toggleModal,
    children: props.children
  }));
};
const SidebarSearchModal = (props) => {
  return /* @__PURE__ */ React.createElement(SearchModalProvider, null, /* @__PURE__ */ React.createElement(SidebarSearchModalContent, {
    ...props
  }));
};

export { SidebarSearchModal };
